### ISSUE TITLE
IWYU fails with "expected exactly one compiler job" error with multiple -arch flags

### ISSUE THREAD

**Issue by scztt** [include-what-you-use/include-what-you-use#1279]
When running iwyu with multiple -arch flags, the tools fails with an error:

```
> include-what-you-use test.cpp -arch arm64 -arch x86_64

error: unable to handle compilation, expected exactly one compiler job in ' "/opt/homebrew/Cellar/include-what-you-use/0.20_1/libexec/bin/include-what-you-use" "-cc1" "-triple" "arm64-apple-macosx12.0.0" "-Wundef-prefix=TARGET_OS_" "-Werror=undef-prefix" "-Wdeprecated-objc-isa-usage" "-Werror=deprecated-objc-isa-usage" "-fsyntax-only" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "test.cpp" "-mrelocation-model" "pic" "-pic-level" "2" "-mframe-pointer=non-leaf" "-ffp-contract=on" "-fno-rounding-math" "-funwind-tables=1" "-fcompatibility-qualified-id-block-type-checking" "-fvisibility-inlines-hidden-static-local-var" "-target-cpu" "apple-m1" "-target-feature" "+v8.5a" "-target-feature" "+crc" "-target-feature" "+crypto" "-target-feature" "+dotprod" "-target-feature" "+fp-armv8" "-target-feature" "+fp16fml" "-target-feature" "+lse" "-target-feature" "+ras" "-target-feature" "+rcpc" "-target-feature" "+rdm" "-target-feature" "+neon" "-target-feature" "+zcm" "-target-feature" "+zcz" "-target-feature" "+fullfp16" "-target-feature" "+sm4" "-target-feature" "+sha3" "-target-feature" "+sha2" "-target-feature" "+aes" "-target-abi" "darwinpcs" "-mllvm" "-treat-scalable-fixed-error-as-warning" "-debugger-tuning=lldb" "-target-linker-version" "820.1" "-fcoverage-compilation-dir=/Users/Shared/_code/native-ui" "-resource-dir" "/opt/homebrew/Cellar/include-what-you-use/0.20_1/libexec/lib/clang/16" "-isysroot" "/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk" "-internal-isystem" "/opt/homebrew/Cellar/include-what-you-use/0.20_1/libexec/bin/../include/c++/v1" "-internal-isystem" "/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/local/include" "-internal-isystem" "/opt/homebrew/Cellar/include-what-you-use/0.20_1/libexec/lib/clang/16/include" "-internal-externc-isystem" "/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include" "-fdeprecated-macro" "-fdebug-compilation-dir=/Users/Shared/_code/native-ui" "-ferror-limit" "19" "-stack-protector" "1" "-fblocks" "-fencode-extended-block-signature" "-fregister-global-dtors-with-atexit" "-fgnuc-version=4.2.1" "-fcxx-exceptions" "-fexceptions" "-fmax-type-align=16" "-D__GCC_HAVE_DWARF2_CFI_ASM=1" "-x" "c++" "test.cpp";  "/opt/homebrew/Cellar/include-what-you-use/0.20_1/libexec/bin/include-what-you-use" "-cc1" "-triple" "x86_64-apple-macosx12.0.0" "-Wundef-prefix=TARGET_OS_" "-Werror=undef-prefix" "-Wdeprecated-objc-isa-usage" "-Werror=deprecated-objc-isa-usage" "-fsyntax-only" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "test.cpp" "-mrelocation-model" "pic" "-pic-level" "2" "-mframe-pointer=all" "-ffp-contract=on" "-fno-rounding-math" "-funwind-tables=2" "-fcompatibility-qualified-id-block-type-checking" "-fvisibility-inlines-hidden-static-local-var" "-target-cpu" "penryn" "-tune-cpu" "generic" "-mllvm" "-treat-scalable-fixed-error-as-warning" "-debugger-tuning=lldb" "-target-linker-version" "820.1" "-fcoverage-compilation-dir=/Users/Shared/_code/native-ui" "-resource-dir" "/opt/homebrew/Cellar/include-what-you-use/0.20_1/libexec/lib/clang/16" "-isysroot" "/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk" "-internal-isystem" "/opt/homebrew/Cellar/include-what-you-use/0.20_1/libexec/bin/../include/c++/v1" "-internal-isystem" "/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/local/include" "-internal-isystem" "/opt/homebrew/Cellar/include-what-you-use/0.20_1/libexec/lib/clang/16/include" "-internal-externc-isystem" "/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include" "-fdeprecated-macro" "-fdebug-compilation-dir=/Users/Shared/_code/native-ui" "-ferror-limit" "19" "-stack-protector" "1" "-fblocks" "-fencode-extended-block-signature" "-fregister-global-dtors-with-atexit" "-fgnuc-version=4.2.1" "-fcxx-exceptions" "-fexceptions" "-fmax-type-align=16" "-D__GCC_HAVE_DWARF2_CFI_ASM=1" "-x" "c++" "test.cpp"; '
```

I imagine fixing this would not be straightforward, as I presume there would need to be one include pass for each arch, since it could potentially alter what is included. My guess is each would need to be checked, and the result merged... somehow.

However, this fundamentally breaks usage for any dual-architecture compiles (I'm not super familiar with the details, but I assume this is standard on mac now). It might be enough to either just choose the first arch, or offer a iwyu flag to e.g. pick which of several archs to actually run the check against? At least with a iwyu flag, a `compile_command.json` that contains multiple `-arch` flags in its command lines is still usable - either is better than the tool simply not working for a common mac build scenario.

**Comment by kimgr** [comment]
I think it's a bit more complicated than "don't care" since arch flags affect the include graph.

IWYU is very much oriented around "one invocation, one output" so it might be hard to make it run one pass for every arch and still get comprehensible and consistent output.

My preference is obviously (🙂) to leave IWYU as-is and rewrite the compilation database to drop unwanted args.

**Comment by scztt** [comment]
> According to godbolt, the compiler doesn't care ( https://godbolt.org/z/9dPME49c1 ), so neither should iwyu.

I think this is because godbolt appends it's own processor arch settings implicitly. It would be a bit baffling if `clang` ignored it's own `-arch` flags, and they definitely matter to iwyu (https://godbolt.org/z/Y8nhzTcr6).

----

> I think it's a bit more complicated than "don't care" since arch flags affect the include graph.

Agreed - probably a good start here would be to specify an error case for this, since it represents a compiler flag case that isn't supported by IWYU? Are there other existing cases where IWYU bails out because of explicitly unsupported compiler flags / build scenarios?

> My preference is obviously (🙂) to leave IWYU as-is and rewrite the compilation database to drop unwanted args.

It's not really a compilation database issue in the end - I could have just as easily provided a set of valid compiler flags directly to iwyu and had the same failure (I was tempted to try a quick fix for this in `iwyu_tool.py` but it's not really the right place). Ultimately it seems like this EITHER requires:
-  a way to have IWYU optionally transform the compile command line before running (by e.g. providing a flag / option to filter for only one arch), and then letting the user piece together any differences between each of the IWYU outputs (where, realistically, these would be identical in the majority of cases) 
- or, letting IWYU split the command line into separate runs (for each arch), doing its analysis on each, and then merging the results (presumably wrapping include differences between each arch in a `#if defined(__x86_64__)` or something).

The latter is probably the correct solution, and also the most challenging option to implement :)

I wonder if there are other clang flags that would implicitly generate multiple compilation runs?

**Comment by kimgr** [comment]
It looks like Clang Tooling has some code that slices out the relevant compilation job: https://github.com/llvm/llvm-project/blob/main/clang/lib/Tooling/Tooling.cpp#L133 (and with special sauce for Mac offload compilation, which I think might be what we're hitting here). We should probably do something similar.

**Comment by kimgr** [comment]
From what I can tell of the Clang Tooling solution above, it picks an arbitrary compile job and uses that. I implemented something similar yesterday, and it seems to work.

The drawback is that we will use whatever "arch" comes first, without affording any control to the user to select a particular architecture when multiple are present. I'm personally fine with that; IWYU is already hampered in the face of preprocessor conditionals (it can only evaluate the path that is actually enabled in the compilation), so I think I'll polish it up for review. Users who want more control will have to rewrite their IWYU commands to only target a single arch, in that case.

While we could add a command-line arg to select, I'd really prefer not to, as arguments come with a lot of commitment for stability, documentation, etc.

**Comment by kimgr** [comment]
Let me know if you can live with https://github.com/include-what-you-use/include-what-you-use/pull/1295 as a fix.

**Comment by kimgr** [comment]
@scztt Ping! Have you had a chance to try PR #1295?

**Comment by kimgr** [comment]
Fixed by filtering out offload jobs in [665b635](https://github.com/include-what-you-use/include-what-you-use/commit/665b63596e625d3be287f447688bd9385f7bad2c).
